File: app/src/main/java/io/github/summerdez/asmrplayer/data/remote/DlsiteLibraryJsonParser.java
package io.github.summerdez.asmrplayer.data.remote;

import io.github.summerdez.asmrplayer.R;
import io.github.summerdez.asmrplayer.data.*;
import io.github.summerdez.asmrplayer.data.remote.*;
import io.github.summerdez.asmrplayer.data.download.*;
import io.github.summerdez.asmrplayer.data.files.*;
import io.github.summerdez.asmrplayer.domain.*;
import io.github.summerdez.asmrplayer.domain.model.*;
import io.github.summerdez.asmrplayer.playback.*;
import io.github.summerdez.asmrplayer.presentation.*;
import io.github.summerdez.asmrplayer.ui.*;
import io.github.summerdez.asmrplayer.ui.activity.*;
import io.github.summerdez.asmrplayer.ui.components.*;
import io.github.summerdez.asmrplayer.ui.screens.*;
import io.github.summerdez.asmrplayer.ui.theme.*;
import io.github.summerdez.asmrplayer.ui.util.*;
import io.github.summerdez.asmrplayer.di.*;
import java.net.URI;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;

final class DlsiteLibraryJsonParser {
    private static final String PLAY_BASE_URL = "https://play.dlsite.com";

    private DlsiteLibraryJsonParser() {
    }

    static boolean isAuthorized(String json) {
        try {
            Object root = DlsiteJsonSupport.parse(json);
            if (root instanceof Map) {
                return !((Map<?, ?>) root).isEmpty();
            }
            if (root instanceof List) {
                return !((List<?>) root).isEmpty();
            }
            return root != null;
        } catch (IllegalArgumentException exception) {
            return false;
        }
    }

    static DlsiteJsonParser.ContentCount parseContentCount(String json)
            throws DlsiteJsonParser.IOExceptionLikeJsonException {
        try {
            Map<String, Object> object = DlsiteJsonSupport.asObject(DlsiteJsonSupport.parse(json));
            return new DlsiteJsonParser.ContentCount(
                    DlsiteJsonSupport.asInt(object.get("user"), 0),
                    DlsiteJsonSupport.asInt(object.get("production"), 0),
                    DlsiteJsonSupport.asInt(object.get("page_limit"), 50));
        } catch (IllegalArgumentException exception) {
            throw new DlsiteJsonParser.IOExceptionLikeJsonException("购买数量解析失败", exception);
        }
    }

    static List<String> parseSalesWorkIds(String json)
            throws DlsiteJsonParser.IOExceptionLikeJsonException {
        try {
            return parseWorkIds(json, "sales");
        } catch (IllegalArgumentException exception) {
            throw new DlsiteJsonParser.IOExceptionLikeJsonException("购买记录解析失败", exception);
        }
    }

    static List<String> parseHistoryWorkIds(String json)
            throws DlsiteJsonParser.IOExceptionLikeJsonException {
        try {
            return parseWorkIds(json, "histories");
        } catch (IllegalArgumentException exception) {
            throw new DlsiteJsonParser.IOExceptionLikeJsonException("浏览记录解析失败", exception);
        }
    }

    static List<DlsiteWork> parseContentWorks(String json)
            throws DlsiteJsonParser.IOExceptionLikeJsonException {
        try {
            List<Object> array = DlsiteJsonSupport.arrayFromRoot(DlsiteJsonSupport.parse(json), "works");
            List<DlsiteWork> works = new ArrayList<>();
            for (Object value : array) {
                DlsiteWork work = toWork(DlsiteJsonSupport.asObjectOrNull(value));
                if (work != null) {
                    works.add(work);
                }
            }
            return works;
        } catch (IllegalArgumentException exception) {
            throw new DlsiteJsonParser.IOExceptionLikeJsonException("作品解析失败", exception);
        }
    }

    static DlsiteWork parseWorkDetail(String json)
            throws DlsiteJsonParser.IOExceptionLikeJsonException {
        try {
            Map<String, Object> root = DlsiteJsonSupport.asObject(DlsiteJsonSupport.parse(json));
            Map<String, Object> work = DlsiteJsonSupport.asObjectOrNull(root.get("work"));
            return toWork(work == null ? root : work);
        } catch (IllegalArgumentException exception) {
            throw new DlsiteJsonParser.IOExceptionLikeJsonException("作品详情解析失败", exception);
        }
    }

    private static List<String> parseWorkIds(String json, String key) {
        List<Object> array = DlsiteJsonSupport.arrayFromRoot(DlsiteJsonSupport.parse(json), key);
        List<String> ids = new ArrayList<>();
        for (Object value : array) {
            Map<String, Object> object = DlsiteJsonSupport.asObjectOrNull(value);
            if (object == null) {
                continue;
            }
            String workId = DlsiteJsonSupport.asString(object.get("workno")).trim();
            if (!workId.isEmpty() && !ids.contains(workId)) {
                ids.add(workId);
            }
        }
        return ids;
    }

    private static DlsiteWork toWork(Map<String, Object> object) {
        if (object == null) {
            return null;
        }
        String workId = DlsiteJsonSupport.asString(object.get("workno")).trim();
        if (workId.isEmpty() || !isDownloadableAudio(object)) {
            return null;
        }
        String title = localizedString(object.get("name"));
        if (title.isEmpty()) {
            title = localizedString(object.get("work_name"));
        }
        if (title.isEmpty()) {
            title = DlsiteJsonSupport.asString(object.get("title")).trim();
        }
        if (title.isEmpty()) {
            title = workId;
        }
        String coverUrl = coverUrlFrom(object);
        if (coverUrl.isEmpty()) {
            coverUrl = DlsiteWork.inferredCoverUrl(workId);
        }
        return new DlsiteWork(
                workId,
                title,
                PLAY_BASE_URL + "/work/" + workId + "/tree",
                PLAY_BASE_URL + "/api/v3/download?workno=" + workId,
                coverUrl);
    }

    private static boolean isDownloadableAudio(Map<String, Object> object) {
        Boolean downloadable = DlsiteJsonSupport.asBooleanOrNull(object.get("downloadable"));
        if (Boolean.FALSE.equals(downloadable)) {
            return false;
        }
        if (object.containsKey("content_count")
                && DlsiteJsonSupport.asInt(object.get("content_count"), 0) <= 0) {
            return false;
        }
        String appType = DlsiteJsonSupport.asString(object.get("app_type")).trim();
        if ("playbox".equalsIgnoreCase(appType)) {
            return false;
        }
        if ("sound".equalsIgnoreCase(appType)) {
            return true;
        }
        String workType = DlsiteJsonSupport.asString(object.get("work_type")).trim();
        String category = DlsiteJsonSupport.asString(object.get("work_category")).trim();
        return "SOU".equalsIgnoreCase(workType)
                || "MUS".equalsIgnoreCase(workType)
                || "music".equalsIgnoreCase(category)
                || "audio".equalsIgnoreCase(category);
    }

    private static String coverUrlFrom(Map<String, Object> object) {
        String[] directKeys = new String[]{
                "coverUrl", "cover_url", "cover", "jacket", "jacket_url",
                "image_main", "imageMain", "imageMainUrl", "image_main_url",
                "main_image", "mainImage", "mainImageUrl", "main_image_url",
                "work_image", "workImage", "workImageUrl", "work_image_url",
                "thumbnail", "thumbnail_url", "thumbnailUrl",
                "thumb", "thumb_url", "thumbUrl",
                "image", "image_url", "imageUrl",
                "images", "image_urls", "imageUrls",
                "sam", "sample_image", "sampleImage"
        };
        for (String key : directKeys) {
            String url = coverUrlCandidate(object.get(key));
            if (!url.isEmpty()) {
                return url;
            }
        }
        for (Map.Entry<String, Object> entry : object.entrySet()) {
            String key = entry.getKey() == null ? "" : entry.getKey().toLowerCase(Locale.US);
            if (!key.contains("cover")
                    && !key.contains("jacket")
                    && !key.contains("image")
                    && !key.contains("thumb")
                    && !key.equals("sam")) {
                continue;
            }
            String url = coverUrlCandidate(entry.getValue());
            if (!url.isEmpty()) {
                return url;
            }
        }
        return "";
    }

    private static String coverUrlCandidate(Object value) {
        if (value instanceof String) {
            return normalizeCoverUrl((String) value);
        }
        List<Object> list = DlsiteJsonSupport.asListOrNull(value);
        if (list != null) {
            for (Object item : list) {
                String url = coverUrlCandidate(item);
                if (!url.isEmpty()) {
                    return url;
                }
            }
            return "";
        }
        Map<String, Object> object = DlsiteJsonSupport.asObjectOrNull(value);
        if (object == null) {
            return "";
        }

        String[] preferredKeys = new String[]{
                "url", "src", "href", "main", "large", "large_url", "largeUrl",
                "medium", "medium_url", "mediumUrl", "original", "original_url", "originalUrl",
                "work", "image", "thumbnail", "thumb", "jacket", "cover",
                "path", "file", "filename", "file_name", "sam"
        };
        for (String key : preferredKeys) {
            String url = coverUrlCandidate(object.get(key));
            if (!url.isEmpty()) {
                return url;
            }
        }
        for (Object nestedValue : object.values()) {
            String url = coverUrlCandidate(nestedValue);
            if (!url.isEmpty()) {
                return url;
            }
        }
        return "";
    }

    private static String normalizeCoverUrl(String value) {
        String url = value == null ? "" : value.trim();
        if (url.isEmpty()) {
            return "";
        }
        String lower = url.toLowerCase(Locale.US);
        if (!lower.startsWith("http://")
                && !lower.startsWith("https://")
                && !lower.startsWith("//")
                && !lower.startsWith("/")) {
            return "";
        }
        try {
            return URI.create(PLAY_BASE_URL).resolve(url).toString();
        } catch (IllegalArgumentException exception) {
            return "";
        }
    }

    private static String localizedString(Object value) {
        Map<String, Object> object = DlsiteJsonSupport.asObjectOrNull(value);
        if (object != null) {
            String[] preferred = new String[]{"zh_CN", "ja_JP", "en_US", "zh_TW", "ko_KR"};
            for (String key : preferred) {
                String text = DlsiteJsonSupport.asString(object.get(key)).trim();
                if (!text.isEmpty()) {
                    return text;
                }
            }
            for (Object nestedValue : object.values()) {
                String text = DlsiteJsonSupport.asString(nestedValue).trim();
                if (!text.isEmpty()) {
                    return text;
                }
            }
            return "";
        }
        return DlsiteJsonSupport.asString(value).trim();
    }
}

File: app/src/main/java/io/github/summerdez/asmrplayer/data/remote/DlsiteLibraryJsonParser.kt
package io.github.summerdez.asmrplayer.data.remote

import io.github.summerdez.asmrplayer.domain.model.DlsiteWork
import java.net.URI
import java.util.Locale

object DlsiteLibraryJsonParser {
    private const val PLAY_BASE_URL = "https://play.dlsite.com"

    fun isAuthorized(json: String): Boolean {
        return try {
            when (val root = DlsiteJsonSupport.parse(json)) {
                is Map<*, *> -> root.isNotEmpty()
                is List<*> -> root.isNotEmpty()
                else -> root != null
            }
        } catch (exception: IllegalArgumentException) {
            false
        }
    }

    @Throws(DlsiteJsonParser.IOExceptionLikeJsonException::class)
    fun parseContentCount(json: String): DlsiteJsonParser.ContentCount {
        return try {
            val data = DlsiteJsonSupport.asObject(DlsiteJsonSupport.parse(json))
            DlsiteJsonParser.ContentCount(
                DlsiteJsonSupport.asInt(data["user"], 0),
                DlsiteJsonSupport.asInt(data["production"], 0),
                DlsiteJsonSupport.asInt(data["page_limit"], 50),
            )
        } catch (exception: IllegalArgumentException) {
            throw DlsiteJsonParser.IOExceptionLikeJsonException("购买数量解析失败", exception)
        }
    }

    @Throws(DlsiteJsonParser.IOExceptionLikeJsonException::class)
    fun parseSalesWorkIds(json: String): List<String> {
        return try {
            parseWorkIds(json, "sales")
        } catch (exception: IllegalArgumentException) {
            throw DlsiteJsonParser.IOExceptionLikeJsonException("购买记录解析失败", exception)
        }
    }

    @Throws(DlsiteJsonParser.IOExceptionLikeJsonException::class)
    fun parseHistoryWorkIds(json: String): List<String> {
        return try {
            parseWorkIds(json, "histories")
        } catch (exception: IllegalArgumentException) {
            throw DlsiteJsonParser.IOExceptionLikeJsonException("浏览记录解析失败", exception)
        }
    }

    @Throws(DlsiteJsonParser.IOExceptionLikeJsonException::class)
    fun parseContentWorks(json: String): List<DlsiteWork> {
        return try {
            val array = DlsiteJsonSupport.arrayFromRoot(DlsiteJsonSupport.parse(json), "works")
            val works = ArrayList<DlsiteWork>()
            for (value in array) {
                val work = toWork(DlsiteJsonSupport.asObjectOrNull(value))
                if (work != null) {
                    works.add(work)
                }
            }
            works
        } catch (exception: IllegalArgumentException) {
            throw DlsiteJsonParser.IOExceptionLikeJsonException("作品解析失败", exception)
        }
    }

    @Throws(DlsiteJsonParser.IOExceptionLikeJsonException::class)
    fun parseWorkDetail(json: String): DlsiteWork? {
        return try {
            val root = DlsiteJsonSupport.asObject(DlsiteJsonSupport.parse(json))
            val work = DlsiteJsonSupport.asObjectOrNull(root["work"])
            toWork(work ?: root)
        } catch (exception: IllegalArgumentException) {
            throw DlsiteJsonParser.IOExceptionLikeJsonException("作品详情解析失败", exception)
        }
    }

    private fun parseWorkIds(json: String, key: String): List<String> {
        val array = DlsiteJsonSupport.arrayFromRoot(DlsiteJsonSupport.parse(json), key)
        val ids = ArrayList<String>()
        for (value in array) {
            val data = DlsiteJsonSupport.asObjectOrNull(value) ?: continue
            val workId = javaTrim(DlsiteJsonSupport.asString(data["workno"]))
            if (workId.isNotEmpty() && !ids.contains(workId)) {
                ids.add(workId)
            }
        }
        return ids
    }

    private fun toWork(data: Map<String, Any?>?): DlsiteWork? {
        if (data == null) {
            return null
        }
        val workId = javaTrim(DlsiteJsonSupport.asString(data["workno"]))
        if (workId.isEmpty() || !isDownloadableAudio(data)) {
            return null
        }
        var title = localizedString(data["name"])
        if (title.isEmpty()) {
            title = localizedString(data["work_name"])
        }
        if (title.isEmpty()) {
            title = javaTrim(DlsiteJsonSupport.asString(data["title"]))
        }
        if (title.isEmpty()) {
            title = workId
        }
        var coverUrl = coverUrlFrom(data)
        if (coverUrl.isEmpty()) {
            coverUrl = DlsiteWork.inferredCoverUrl(workId)
        }
        return DlsiteWork(
            workId,
            title,
            "$PLAY_BASE_URL/work/$workId/tree",
            "$PLAY_BASE_URL/api/v3/download?workno=$workId",
            coverUrl,
        )
    }

    private fun isDownloadableAudio(data: Map<String, Any?>): Boolean {
        val downloadable = DlsiteJsonSupport.asBooleanOrNull(data["downloadable"])
        if (downloadable == false) {
            return false
        }
        if (data.containsKey("content_count") && DlsiteJsonSupport.asInt(data["content_count"], 0) <= 0) {
            return false
        }
        val appType = javaTrim(DlsiteJsonSupport.asString(data["app_type"]))
        if ("playbox".equals(appType, ignoreCase = true)) {
            return false
        }
        if ("sound".equals(appType, ignoreCase = true)) {
            return true
        }
        val workType = javaTrim(DlsiteJsonSupport.asString(data["work_type"]))
        val category = javaTrim(DlsiteJsonSupport.asString(data["work_category"]))
        return "SOU".equals(workType, ignoreCase = true) ||
            "MUS".equals(workType, ignoreCase = true) ||
            "music".equals(category, ignoreCase = true) ||
            "audio".equals(category, ignoreCase = true)
    }

    private fun coverUrlFrom(data: Map<String, Any?>): String {
        val directKeys = arrayOf(
            "coverUrl", "cover_url", "cover", "jacket", "jacket_url",
            "image_main", "imageMain", "imageMainUrl", "image_main_url",
            "main_image", "mainImage", "mainImageUrl", "main_image_url",
            "work_image", "workImage", "workImageUrl", "work_image_url",
            "thumbnail", "thumbnail_url", "thumbnailUrl",
            "thumb", "thumb_url", "thumbUrl",
            "image", "image_url", "imageUrl",
            "images", "image_urls", "imageUrls",
            "sam", "sample_image", "sampleImage",
        )
        for (key in directKeys) {
            val url = coverUrlCandidate(data[key])
            if (url.isNotEmpty()) {
                return url
            }
        }
        for ((entryKey, entryValue) in data) {
            val key = entryKey.lowercase(Locale.US)
            if (!key.contains("cover") &&
                !key.contains("jacket") &&
                !key.contains("image") &&
                !key.contains("thumb") &&
                key != "sam"
            ) {
                continue
            }
            val url = coverUrlCandidate(entryValue)
            if (url.isNotEmpty()) {
                return url
            }
        }
        return ""
    }

    private fun coverUrlCandidate(value: Any?): String {
        if (value is String) {
            return normalizeCoverUrl(value)
        }
        val list = DlsiteJsonSupport.asListOrNull(value)
        if (list != null) {
            for (item in list) {
                val url = coverUrlCandidate(item)
                if (url.isNotEmpty()) {
                    return url
                }
            }
            return ""
        }
        val data = DlsiteJsonSupport.asObjectOrNull(value) ?: return ""

        val preferredKeys = arrayOf(
            "url", "src", "href", "main", "large", "large_url", "largeUrl",
            "medium", "medium_url", "mediumUrl", "original", "original_url", "originalUrl",
            "work", "image", "thumbnail", "thumb", "jacket", "cover",
            "path", "file", "filename", "file_name", "sam",
        )
        for (key in preferredKeys) {
            val url = coverUrlCandidate(data[key])
            if (url.isNotEmpty()) {
                return url
            }
        }
        for (nestedValue in data.values) {
            val url = coverUrlCandidate(nestedValue)
            if (url.isNotEmpty()) {
                return url
            }
        }
        return ""
    }

    private fun normalizeCoverUrl(value: String?): String {
        val url = javaTrim(value)
        if (url.isEmpty()) {
            return ""
        }
        val lower = url.lowercase(Locale.US)
        if (!lower.startsWith("http://") &&
            !lower.startsWith("https://") &&
            !lower.startsWith("//") &&
            !lower.startsWith("/")
        ) {
            return ""
        }
        return try {
            URI.create(PLAY_BASE_URL).resolve(url).toString()
        } catch (exception: IllegalArgumentException) {
            ""
        }
    }

    private fun localizedString(value: Any?): String {
        val data = DlsiteJsonSupport.asObjectOrNull(value)
        if (data != null) {
            val preferred = arrayOf("zh_CN", "ja_JP", "en_US", "zh_TW", "ko_KR")
            for (key in preferred) {
                val text = javaTrim(DlsiteJsonSupport.asString(data[key]))
                if (text.isNotEmpty()) {
                    return text
                }
            }
            for (nestedValue in data.values) {
                val text = javaTrim(DlsiteJsonSupport.asString(nestedValue))
                if (text.isNotEmpty()) {
                    return text
                }
            }
            return ""
        }
        return javaTrim(DlsiteJsonSupport.asString(value))
    }

    private fun javaTrim(value: String?): String = value?.trim { it <= ' ' }.orEmpty()
}
